core/common/src/main/java/alluxio/collections/LockCache.java

```diff
@@ -0,0 +1,248 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.collections;
+
+import alluxio.resource.LockResource;
+import alluxio.resource.RefCountLockResource;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Function;
+
+/**
+ * A cache specifically designed to contain locks and will NOT evict any entries
+ * that are in use.
+ *
+ * @param <K> key to the cache
+ */
+public class LockCache<K> {
+  /**
+   * Node containing value and other information to be stored in the cache.
+   */
+  public class ValNode {
+    private ReentrantReadWriteLock mValue;
+    private boolean mIsAccessed;
+    private AtomicInteger mRefCount;
+
+    private ValNode(ReentrantReadWriteLock value) {
+      mValue = value;
+      mIsAccessed = false;
+      mRefCount = new AtomicInteger(1);
+    }
+
+    /**
+     * Get the ref counter associated with this value node.
+     *
+     * @return a ref counter
+     */
+    public AtomicInteger getRefCounter() {
+      return mRefCount;
+    }
+
+    /**
+     * Get the value contained in the value node.
+     *
+     * @return the lock contained in the value node
+     */
+    public ReentrantReadWriteLock get() {
+      return mValue;
+    }
+  }
+
+  private static final Logger LOG = LoggerFactory.getLogger(LockCache.class);
+  private static final float DEFAULT_LOAD_FACTOR = 0.75f;
+  static final float SOFT_LIMIT_RATIO = 0.9f;
+
+  private final Map<K, ValNode> mCache;
+
+  /**
+   * SoftLimit = hardLimit * softLimitRatio
+   * once the size reaches softlimit, eviction starts to happen,
+   * once the size reaches hardlimit, blocking starts to happen.
+   */
+  private final int mHardLimit;
+  private final int mSoftLimit;
+  private final Function<? super K, ? extends ReentrantReadWriteLock> mDefaultLoader;
+  private Iterator<Map.Entry<K, ValNode>> mIterator;
+  private final Lock mEvictLock;
+
+  private long mLastLogTime;
```

Comment:
perhaps `mLastSizeWarningTime` to be more specific? We might add more unrelated log messages